File: Sources/Model/ObjC/TracksLogging.m
#import "TracksLogging.h"
@import Foundation;
@import ObjectiveC;


Class<TracksLoggingConfiguration> TracksLoggingClass() {
    static Class<TracksLoggingConfiguration> result = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
#if SWIFT_PACKAGE
        Class loggingClass = NSClassFromString(@"AutomatticTracksModel.TracksLogging");
#else
        Class loggingClass = NSClassFromString(@"TracksLogging");
#endif
        if ([loggingClass conformsToProtocol: @protocol(TracksLoggingConfiguration)]) {
            result = loggingClass;
        }
    });
    return result;
}


void TracksLogError(NSString *format, ...) {
    va_list args;
    va_start(args, format);

    NSString *str = [[NSString alloc] initWithFormat:format arguments:args];
    [[TracksLoggingClass() delegate] logError:str];

    va_end(args);
}

void TracksLogWarn(NSString *format, ...) {
    va_list args;
    va_start(args, format);

    NSString *str = [[NSString alloc] initWithFormat:format arguments:args];
    [[TracksLoggingClass() delegate] logWarning:str];

    va_end(args);
}

void TracksLogInfo(NSString *format, ...) {
    va_list args;
    va_start(args, format);

    NSString *str = [[NSString alloc] initWithFormat:format arguments:args];
    [[TracksLoggingClass() delegate] logInfo:str];

    va_end(args);
}

void TracksLogDebug(NSString *format, ...) {
    va_list args;
    va_start(args, format);

    NSString *str = [[NSString alloc] initWithFormat:format arguments:args];
    [[TracksLoggingClass() delegate] logDebug:str];

    va_end(args);
}

void TracksLogVerbose(NSString *format, ...) {
    va_list args;
    va_start(args, format);

    NSString *str = [[NSString alloc] initWithFormat:format arguments:args];
    [[TracksLoggingClass() delegate] logVerbose:str];

    va_end(args);
}

File: Sources/Model/Swift/TracksLogging.swift
import Foundation

#if SWIFT_PACKAGE
import AutomatticTracksModelObjC
#endif

// The Swift versions of these functions do not accept format strings
// and arguments; use string interpolation instead
public func TracksLogError(_ str: String) {
    TracksLogging.delegate?.logError(str)
}

public func TracksLogWarn(_ str: String) {
    TracksLogging.delegate?.logWarning(str)
}

public func TracksLogInfo(_ str: String) {
    TracksLogging.delegate?.logInfo(str)
}

public func TracksLogDebug(_ str: String) {
    TracksLogging.delegate?.logDebug(str)
}

public func TracksLogVerbose(_ str: String) {
    TracksLogging.delegate?.logVerbose(str)
}



public class TracksLogging: NSObject, TracksLoggingConfiguration {

    private static var _delegate: TracksLoggingDelegate?

    @objc public static var delegate: TracksLoggingDelegate? {
        get {
            #if DEBUG
            if _delegate == nil {
                print("*** A Tracks logging delegate has not been configured, so Tracks will log to the console. If you would like to redirect logging to an existing logging system, assign a logging delegate to TracksLogging.delegate. (You may need to import AutomatticTracksModel, depending on what else you have imported.)")
                _delegate = DefaultTracksLogging()
            }
            #endif

            return _delegate
        }
        set {
            _delegate = newValue
        }
    }
}

private class DefaultTracksLogging: NSObject, TracksLoggingDelegate {

    func logPanic(_ str: String) {
        print("Tracks <☠️>: \(str)")
    }

    func logError(_ str: String) {
        print("Tracks <E>: \(str)")
    }

    func logWarning(_ str: String) {
        print("Tracks <W>: \(str)")
    }

    func logInfo(_ str: String) {
        print("Tracks <I>: \(str)")
    }

    func logDebug(_ str: String) {
        print("Tracks <D>: \(str)")
    }

    func logVerbose(_ str: String) {
        print("Tracks <V>: \(str)")
    }
}
